### DISCUSSION TITLE
Self Learning Topic Publish

### DISCUSSION THREAD

**Discussion by bdsoha** [skyvar/.github#2] [Announcements]
The previous assignment focused on *coding*, and *self-learning*.

This assignment is focused in *workflow*, *best-practices*, and *tooling*.

Have a look at https://github.com/skyvar/self-learning for more details.

**Comment by bdsoha** [comment]
@esterlev23 @EsterWeiss @RivkaBerlinski @RivkaZishinski @MiriamRotenshtein @chananaimen @chanaSlumianski @FridaPshitik @SaraLitman @ShaindyBorzikowsky 

We are almost at the end of the 2nd day of this assignment.
Tomorrow *(22/03/2023)* will be the due date for the assignment to leave Thursday free for review.

Please add a *reaction* to this post to verify that you read it.

**Comment by FridaPshitik** [comment]
we saw the mail and hope finish
________________________________
מאת: Dov Benyomin Sohacheski ***@***.***>
‏‏נשלח: יום שלישי 21 מרץ 2023 15:17
‏‏אל: skyvar/.github ***@***.***>
עותק: F. Pshitik ***@***.***>; Mention ***@***.***>
‏‏נושא: Re: [skyvar/.github] Self Learning Topic Publish (Discussion #2)


@esterlev23<https://github.com/esterlev23> @EsterWeiss<https://github.com/EsterWeiss> @RivkaBerlinski<https://github.com/RivkaBerlinski> @RivkaZishinski<https://github.com/RivkaZishinski> @MiriamRotenshtein<https://github.com/MiriamRotenshtein> @chananaimen<https://github.com/chananaimen> @chanaSlumianski<https://github.com/chanaSlumianski> @FridaPshitik<https://github.com/FridaPshitik> @SaraLitman<https://github.com/SaraLitman> @ShaindyBorzikowsky<https://github.com/ShaindyBorzikowsky>

We are almost at the end of the 2nd day of this assignment.
Tomorrow (22/03/2023) will be the due date for the assignment to leave Thursday free for review.

Please add a reaction to this post to verify that you read it.

**Reply by bdsoha** [reply]
Please use a *reaction*.

**Comment by bdsoha** [comment]
@skyvar/everyone  Once all PRs have been published we can move on to the next of 2 steps.

**Comment by bdsoha** [comment]
@skyvar/everyone Please apply the following 2 tasks to you assignments:

1. Execute any markdown *linter* to reformat your documents.
2. Execute a [spell-check tool](https://github.com/codespell-project/codespell) to fix spelling of your documents.

**Comment by bdsoha** [comment]
@skyvar/everyone Once the abovementioned has been completed, each member will be assigned another teams PR.
You are to review your colleagues PR and provide a **minimum** of 10 suggestions.

The suggestions can be in regard to:

- Verification *(does the document answer the requirements)*.
- Styling.
- Clarity *(was the document understood to someone without background information)*.
- Improvements.
- Mistakes.
- Etc.

Please use this opportunity to utilize the tools provided by GitHub as part of the PR.

**Comment by bdsoha** [comment]
# Conclusions

Overall, a lot of work was invested in the current task. The main purpose of this *(and the previous)* assignment was to test your self-learning skills. The selected topics were also chosen in order to add more tools to your toolkit for future use.

The upcoming sections outline a few areas *(noticed during this assignment)*, which can benefit from improvement.

## Focus on *How*

The majority of tasks were focus on *how* to do or accomplish something. Complete step-by-step guides, *”Click here ➡ type this ➡ click on that …”*. As such, the assignment was regarded as more or less a *technical document* instead of research.

When researching anything, the focus should instead be on *why*, *what*, and *when*. For example, to name a few:

-	*Why* is a given tool `x` used instead of `y`.
-	*Why* we shouldn’t be doing `x`.
-	*What* is happening behind the scenes when we choose `x`.
-	*What* compromises are made when implementing `x`.
-	*What* are the maintenance benefits of `x`.
-	For *what* size team is `x` no longer a valid option.
-	*When* can I use `x` in a given environment `y`.
-	*When* is the cost of `x` justified.
-	*When* will using `x` require additional resources. 

Reading and understanding documentation for a given tool, platform, or service is important, but does not need to be summarized as part of an assignment. However, the decisions and thought process made while choosing the *correct* tool, platform, or service – while considering the current environment and implementation – should be understood and summarized.

Apply a self-evaluation, now that you have completed the assignment, can you answer the question:

> *When* should I use `git rebase` instead of `git merge`?

> *What* is provided by Jira that is not provided by GitHub’s project management?

> *Why* is markdown so popular?

> *When* should I install a service on my computer?

> *What* is considered a best practice for branching when collaborating on a single project?

## Iterative Development

The process in which something is developed or designed is just as crucial *(if not even more crucial)* that the final outcome. Sharing the process with others *(especially team-leaders)*, will make sure that the requirement was properly understood and the implementation well defined. Potential defects are spotted and dealt with early, and the actual progress is considered measurable.

Submitting any form of *work* only when you consider it *finished*, does not allow any room for feedback or suggestions. In many cases, entire tasks need to be rewritten. Sometimes, rephrasing a requirement in your own words can verify it was well understood.

## Pay Attention to Detail

**Any explanation for this section is superfluous.**

It should be considered obvious that when instructed to do exactly `x`, then exactly `x` should be done. In the future, we will learn about automated *tools* that guarantee we conform to:

-	Best practices.
-	Naming conventions.
-	Coding styles.
-	Design patterns.
-	Design principles.
-	Workflows.
-	Communication.
-	Etc.

It is a shame for a developer to be penalized over the inattentiveness and disregard to details. 

In the open-source community, code is contributed by thousands of develops each with their own coding style and workflow. Nonetheless, when working a given project, they are forced to conform to that project’s workflow and coding style. This presents the illusion that the entire codebase was developed by a single developer, when in fact it was a joint contribution of many.

## Free Time

If you find yourself asking:

> I have completed my task, what should I do in the meantime? 

First and foremost, review your code. Even if you feel it is complete and there is no room for improvement, **review anyway**.

Secondly, utilize your free time to learn new topics or delve deeper into topics you already know. The difference between a *good* developer and a *great* developer is knowledge, experience, and skills.

Prepare a list on the side of ideas or concepts you want to research in your free time. There are so many fields within *software development*, there is no reason to sit around waiting for a new allocated task.

## Code Review

Code review is a very important topic and will be touched upon a lot in the future. For now, it is important to remember the following:

- **As a reviewer:** try to be positive and give encouraging feedback and suggestions. In some cases, a maintainer *(not a contributor)*, can be more critical when providing feedback. Nonetheless, never scrutinize. 
    > **Note**: I left the 👎 reaction to try and stimulate and force the contributors to **think twice** before publishing code. Such practices should not be considered at any point, they can be destructive to a healthy working environment.
- **As a reviewee:** do not take feedback personally. Comments and suggestions left by others are usually constructive and meant as mutual learning for you as well as the reviewer. Input from your peers is crucial, as your code may also affect their outcome.

**Comment by FridaPshitik** [comment]
Should the documentation file be written in a markdown or txt file ?

**Reply by bdsoha** [reply]
Markdown is always best practice.

**Comment by FridaPshitik** [comment]
which points the document suppose to contain ?

**Reply by bdsoha** [reply]
What document are you referring to?
The [post above](https://github.com/orgs/skyvar/discussions/2#discussioncomment-5430478) is a conclusion to the previous task.

**Reply by FridaPshitik** [reply]
the documentation file.

**Reply by bdsoha** [reply]
I am still missing what you are referring to. 
The assignment was to *research* a given topic and provide a detailed overview, not a *user guide*.
